### PR TITLE
Optimizations to ecrecover

### PR DESCRIPTION
**Pull request by hecmas**
This PR introduces:

- Changes the usage of `ARITH` to `ARITH_MOD` to avoid post alias checks in the former.
- Since now functions over the base/scalar field of the curve are 3 lines large, most of them have been deleted and introduced directly in `ecrecover` to avoid unnecessary steps in `CALL/RETURN`s.
- Some minor fixes/optimizations found along the way have been applied.

### CODE REVIEW

File: main/ecrecover/FNSECP256K1/mulFnSecp256k1.zkasm [hunk 1]
```diff
@@ -9,20 +9,13 @@
...
 
-        ; 1] Compute and check the multiplication over Z
-        ; A·B + 0 = [D]·2²⁵⁶ + [E]
+        ; A·B + 0 = op (mod SECP256K1_N)
         0 => C
-        $${var _mulFnSecp256k1_AB = A * B}
-        ${_mulFnSecp256k1_AB >> 256} => D
-        ${_mulFnSecp256k1_AB} => E :ARITH
+        %SECP256K1_N => D
+        ${(A*B + C) % D} => E   :ARITH_MOD
```
Comment (zkronos73) [inline main/ecrecover/FNSECP256K1/mulFnSecp256k1.zkasm:19]: optimization (-1 step):
; C changes in row after ARITH, remove C addition from executor.
${(A*B) % D} => C :ARITH_MOD, RETURN

File: main/ecrecover/FPSECP256K1/addFpSecp256k1.zkasm [hunk 2]
```diff
@@ -9,21 +9,13 @@
...
 
-        ; 1] Compute and check the sum over Z
-        ; A·[1] + C = [D]·2²⁵⁶ + [E]
+        ; A·1 + C = op (mod SECP256K1_P)
         1 => B
-        $${var _addFpSecp256k1_AC = A + C}
-        ${_addFpSecp256k1_AC >> 256} => D
-        ${_addFpSecp256k1_AC} => E :ARITH
+        %SECP256K1_P => D
+        ${(A*B + C) % D} => E   :ARITH_MOD
```
Comment (zkronos73) [inline main/ecrecover/FPSECP256K1/addFpSecp256k1.zkasm:19]: optimization (-1 step):
; C changes in row after ARITH and executor no need to calculate A*B,
${(A + C) % D} => C :ARITH_MOD, RETURN

File: main/ecrecover/FPSECP256K1/mulFpSecp256k1.zkasm [hunk 3]
```diff
@@ -9,20 +9,13 @@
...
 
-        ; 1] Compute and check the multiplication over Z
-        ; A·B + 0 = [D]·2²⁵⁶ + [E]
+        ; A·B + 0 = op (mod SECP256K1_P)
         0 => C
-        $${var _mulFpSecp256k1_AB = A * B}
-        ${_mulFpSecp256k1_AB >> 256} => D
-        ${_mulFpSecp256k1_AB} => E    :ARITH
+        %SECP256K1_P => D
+        ${(A*B + C) % D} => E   :ARITH_MOD
```
Comment (zkronos73) [inline main/ecrecover/FPSECP256K1/mulFpSecp256k1.zkasm:19]: optimization (-1 step):
; C changes in row after ARITH
${(A*B) % D} => C :ARITH_MOD, RETURN

File: main/ecrecover/FPSECP256K1/checkSqrtFpSecp256k1.zkasm [hunk 4]
```diff
@@ -16,20 +16,20 @@
...
 
 checkSqrtFpSecp256k1:
         ; RESOURCES:
-        ;   248 x (3 steps + 2 arith + 8 steps + 2 arith + 7 steps) = 992 ariths + 4464 steps
-        ;   6 x (1 step + 2 arith + 8 steps) = 12 ariths + 54 steps
+        ;   248 x (3 steps + 1 arith + 5 steps + 1 arith + 4 steps) = 496 ariths + 2976 steps
+        ;   6 x (1 step + 1 arith + 5 steps) = 6 ariths + 36 steps
         ;   6 steps + 1 binary
-        ; TOTAL = 1004 ariths + 1 binary + 4524 steps
+        ; TOTAL = 502 ariths + 1 binary + 3018 steps
```
Comment (zkronos73) [inline main/ecrecover/FPSECP256K1/checkSqrtFpSecp256k1.zkasm:32]: OPTIMIZATION, how now the operations are very short, unroll the operations or alternative, reduce at least setting of  SECP256K1_P, call and you can left result in best register for next operation.

File: main/ecrecover/FPSECP256K1/squareFpSecp256k1.zkasm [hunk 5]
```diff
@@ -9,22 +9,14 @@
...
-        ; A·A + 0 = [D]·2²⁵⁶ + [E]
+        ; C·C + 0 = op (mod SECP256K1_P)
         C => A,B
-        0n => C
-        $${var _squareFpSecp256k1_AA = A * A}
-        ${_squareFpSecp256k1_AA >> 256} => D
-        ${_squareFpSecp256k1_AA} => E :ARITH
+        0 => C
+        %SECP256K1_P => D
+        ${(A*B + C) % D} => E   :ARITH_MOD
```
Comment (zkronos73) [inline main/ecrecover/FPSECP256K1/squareFpSecp256k1.zkasm:20]: same optimization as previous:
${(A*B) % D} => C :ARITH_MOD

**Review by zkronos73 — COMMENTED** [review]
some optimizations, not bugs.
changes of utils.zkasm no sense inside this PR, perhaps need to do a rebase?

File: main/ecrecover/FNSECP256K1/mulFnSecp256k1.zkasm (hunk 1 above)
Comment (hecmas) [reply main/ecrecover/FNSECP256K1/mulFnSecp256k1.zkasm:19]: Solved

File: main/ecrecover/FPSECP256K1/addFpSecp256k1.zkasm (hunk 2 above)
Comment (hecmas) [reply main/ecrecover/FPSECP256K1/addFpSecp256k1.zkasm:19]: Added

File: main/ecrecover/FPSECP256K1/mulFpSecp256k1.zkasm (hunk 3 above)
Comment (hecmas) [reply main/ecrecover/FPSECP256K1/mulFpSecp256k1.zkasm:19]: Added

File: main/ecrecover/FPSECP256K1/squareFpSecp256k1.zkasm (hunk 5 above)
Comment (hecmas) [reply main/ecrecover/FPSECP256K1/squareFpSecp256k1.zkasm:20]: Added

File: main/ecrecover/FPSECP256K1/checkSqrtFpSecp256k1.zkasm (hunk 4 above)
Comment (hecmas) [reply main/ecrecover/FPSECP256K1/checkSqrtFpSecp256k1.zkasm]: Done

**Review by krlosMata — APPROVED** [review]
:smile_cat:

**Pull request merged**